### PR TITLE
Color selector for well layer

### PR DESCRIPTION
**Pull request by shruthirai**

### CODE REVIEW

**Comment by shadab-skhan** [comment]
Tried clicking on color scale, the color selector doesn't pops up

**Comment by shruthirai** [comment]
> @shruthirai Tried clicking on color scale, the color selector doesn't pops up

@shadab-skhan Currently it will work only for storybook example of welllayer and not default.

**Review by hkfb — CHANGES_REQUESTED** [review]
When I click to expand or contract any of the menus in the color selector, then the zoom in the map view is reset.

Also, when the color selector is first shown, then I can't find a way to hide it again.

Also, the Custom Color Scale doesn't do anything, so it should be removed or hidden until it is useful.

File: react/src/lib/components/DeckGLMap/DeckGLMap.tsx [hunk 1]
```diff
@@ -43,6 +43,7 @@ export interface DeckGLMapProps {
      */
     checkDatafileSchema?: boolean;
     onMouseEvent?: (event: MapMouseEvent) => void;
+    colorMapping?: (t: number) => string | string[];
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/DeckGLMap.tsx:46]: This should be a layer prop. We need to be able to set separate color mapping to each layer.

**Review by hkfb — CHANGES_REQUESTED** [review]

**Comment by shruthirai** [comment]
> When I click to expand or contract any of the menus in the color selector, then the zoom in the map view is reset.
> 
> Also, when the color selector is first shown, then I can't find a way to hide it again.
> Clicking upon the legend will hide or show (toggle) the color selector. 
>
> Also, the Custom Color Scale doesn't do anything, so it should be removed or hidden until it is useful.

**Comment by hkfb** [comment]
I get the following error when clicking on the color legend.

```
Warning: Each child in a list should have a unique "key" prop.

Check the render method of `ColorSelectorWrapper`. See https://reactjs.org/link/warning-keys for more information.
    at div
    at ColorSelectorWrapper (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35026:34)
    at div
    at O (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:19741)
    at AccordionPanel (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36510:5)
    at div
    at AccordionItem (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36417:5)
    at div
    at Fe (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:17300)
    at Accordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35999:5)
    at div
    at O (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:19741)
    at AccordionPanel (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36510:5)
    at div
    at AccordionItem (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36417:5)
    at div
    at Fe (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:17300)
    at Accordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35999:5)
    at div
    at ColorSelectorAccordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:34673:24)
    at div
    at div
    at ColorLegend (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35080:23)
    at div
    at div
    at Provider (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:459697:20)
    at unboundStoryFn (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:222679:12)
    at ErrorBoundary (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:220103:5)
printWarning @ react.development.jss:220
error @ react.development.jss:196
validateExplicitKey @ react.development.jss:2052
validateChildKeys @ react.development.jss:2078
createElementWithValidation @ react.development.jss:2233
ColorSelectorWrapper @ ColorSelectorWrapper.jss:61
renderWithHooks @ react-dom.development.jss:14985
mountIndeterminateComponent @ react-dom.development.jss:17811
beginWork @ react-dom.development.jss:19049
beginWork$1 @ react-dom.development.jss:23940
performUnitOfWork @ react-dom.development.jss:22779
workLoopSync @ react-dom.development.jss:22707
renderRootSync @ react-dom.development.jss:22670
performSyncWorkOnRoot @ react-dom.development.jss:22293
(anonymous) @ react-dom.development.jss:11327
unstable_runWithPriority @ scheduler.development.jss:468
runWithPriority$1 @ react-dom.development.jss:11276
flushSyncCallbackQueueImpl @ react-dom.development.jss:11322
flushSyncCallbackQueue @ react-dom.development.jss:11309
discreteUpdates$1 @ react-dom.development.jss:22420
discreteUpdates @ react-dom.development.jss:3756
dispatchDiscreteEvent @ react-dom.development.jss:5889
```

Also, the title of the legend is "Legend", and it looks like it should be "ZONELOG", so that we know which data uses it.

**Comment by shruthirai** [comment]
> 



> I get the following error when clicking on the color legend.
> 
> ```
> Warning: Each child in a list should have a unique "key" prop.
> 
> Check the render method of `ColorSelectorWrapper`. See https://reactjs.org/link/warning-keys for more information.
>     at div
>     at ColorSelectorWrapper (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35026:34)
>     at div
>     at O (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:19741)
>     at AccordionPanel (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36510:5)
>     at div
>     at AccordionItem (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36417:5)
>     at div
>     at Fe (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:17300)
>     at Accordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35999:5)
>     at div
>     at O (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:19741)
>     at AccordionPanel (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36510:5)
>     at div
>     at AccordionItem (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:36417:5)
>     at div
>     at Fe (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:479967:17300)
>     at Accordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35999:5)
>     at div
>     at ColorSelectorAccordion (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:34673:24)
>     at div
>     at div
>     at ColorLegend (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:35080:23)
>     at div
>     at div
>     at Provider (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:459697:20)
>     at unboundStoryFn (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:222679:12)
>     at ErrorBoundary (http://localhost:6006/vendors-node_modules_deck_gl_aggregation-layers_dist_esm_index_js-node_modules_deck_gl_extens-fc7d57.iframe.bundle.js:220103:5)
> printWarning @ react.development.jss:220
> error @ react.development.jss:196
> validateExplicitKey @ react.development.jss:2052
> validateChildKeys @ react.development.jss:2078
> createElementWithValidation @ react.development.jss:2233
> ColorSelectorWrapper @ ColorSelectorWrapper.jss:61
> renderWithHooks @ react-dom.development.jss:14985
> mountIndeterminateComponent @ react-dom.development.jss:17811
> beginWork @ react-dom.development.jss:19049
> beginWork$1 @ react-dom.development.jss:23940
> performUnitOfWork @ react-dom.development.jss:22779
> workLoopSync @ react-dom.development.jss:22707
> renderRootSync @ react-dom.development.jss:22670
> performSyncWorkOnRoot @ react-dom.development.jss:22293
> (anonymous) @ react-dom.development.jss:11327
> unstable_runWithPriority @ scheduler.development.jss:468
> runWithPriority$1 @ react-dom.development.jss:11276
> flushSyncCallbackQueueImpl @ react-dom.development.jss:11322
> flushSyncCallbackQueue @ react-dom.development.jss:11309
> discreteUpdates$1 @ react-dom.development.jss:22420
> discreteUpdates @ react-dom.development.jss:3756
> dispatchDiscreteEvent @ react-dom.development.jss:5889
> ```
> 
> Also, the title of the legend is "Legend", and it looks like it should be "ZONELOG", so that we know which data uses it.

Resolved the issue.

File: react/src/lib/components/DeckGLMap/layers/map/mapLayer.ts [hunk 2]
```diff
@@ -10,7 +10,8 @@ import { TerrainLoader } from "@loaders.gl/terrain";
 // XXX import { ImageLoader } from "@loaders.gl/images";
 import { load } from "@loaders.gl/core";
 import { Vector3 } from "@math.gl/core";
-import { colorTablesArray, rgbValues } from "@emerson-eps/color-tables/";
+import { rgbValues } from "@emerson-eps/color-tables";
+import { colorTablesArray } from "@emerson-eps/color-tables/";
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/layers/map/mapLayer.ts:14]: What is the purpose of this change?

File: react/src/lib/components/DeckGLMap/storybook/DeckGLMap.stories.jsx [hunk 3]
```diff
@@ -695,3 +698,99 @@ MultiColorMap.args = {
         ],
     },
 };
+
+// colorselector for welllayer
+const wellLayers = [
+    {
+        "@@type": "WellsLayer",
+        data: "@@#resources.wellsData",
+        logColor: [exampleData[0].layers[4]][0].logColor,
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/storybook/DeckGLMap.stories.jsx:707]: Please avoid using the exampleData json object. It is not a common use case to use json to specify the props, so it distracts from the purpose of the story. Please take a look at the Volve wells example (https://equinor.github.io/webviz-subsurface-components/storybook-static/?path=/story/deckglmap-wells-layer--volve-wells), where the props are defined as js objects. It makes the example easier to understand.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: react/src/lib/components/DeckGLMap/storybook/DeckGLMap.stories.jsx [hunk 4]
```diff
@@ -695,3 +698,99 @@ MultiColorMap.args = {
...
+                <ColorLegend {...args} getColorMapname={wellLayerData} />
+            </div>
+            <DeckGLMap {...args} layers={layers} />
+        </div>
+    );
+};
+
+export const ColorSelectorForWellLayer = wellLayerTemplate.bind({});
+
+ColorSelectorForWellLayer.args = {
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/storybook/DeckGLMap.stories.jsx:781]: Can we put this example in the wells section instead, ie. in `react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.stories.tsx`

**Review by hkfb — CHANGES_REQUESTED** [review]

File: react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.stories.tsx [hunk 5]
```diff
@@ -320,3 +323,99 @@ export const ContinuousColorTable: React.FC = () => {
...
+            >
+                <ColorLegend {...args} getColorMapname={wellLayerData} />
+            </div>
+            <DeckGLMap {...args} layers={layers} />
+        </div>
+    );
+};
+
+//eslint-disable-next-line
+export const WellLayerColorSelector: any = wellLayerTemplate.bind({});
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.stories.tsx:404]: The name is not consistent with other stories. The "Well Layer" prefix is implicit.

File: react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.stories.tsx [hunk 6]
```diff
@@ -320,3 +323,99 @@ export const ContinuousColorTable: React.FC = () => {
...
+
+WellLayerColorSelector.args = {
```
Comment (hkfb) [inline react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.stories.tsx:406]: The example is too small in the Docs page: 

Screenshot from 2022-06-14 10-12-36

I think it would also be useful with a description that lets you know how to change the color scale.

**Review by hkfb — CHANGES_REQUESTED** [review]

**Review by hkfb — APPROVED** [review]
